nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -1098,16 +1076,97 @@ impl<'a> BlueprintBuilder<'a> {
+    ///   decommission it, even if it has had its disposition set to expunged in
+    ///   *this* planner pass.
+    ///
+    ///   However, if the sled is still in service, we must wait for any
+    ///   disks expunged in this planner pass to be seen by the sled-agent and
+    ///   reflected in inventory before we can decommission them. Since any
+    ///   modifications made in this planner pass have not yet been sent to the
+    ///   sled-agents, we only check for expunged sleds in the parent_blueprint.
+    ///
+    ///
+    /// Called by the planner in the `do_plan_decommission()` stage of planning
+    pub fn sled_decommision_disks(
+        &mut self,
+        sled_id: SledUuid,
+        sled_details: &SledDetails,
+    ) -> Result<SledEditCounts, Error> {
+        let editor = self.sled_editors.get_mut(&sled_id).ok_or_else(|| {
+            Error::Planner(anyhow!(
+                "tried to decommission disks for unknown sled {sled_id}"
+            ))
+        })?;
+
+        let initial_counts = editor.edit_counts();
+
+        // This sled is expunged. Decommission all expunged disks. Note that
+        // because the decommision step comes after the expunge step, all disks
+        // should be marked expunged in the editor currently and can therefore
+        // be set to decommissioned.
+        //
+        // We shouldn't have added a disk during this planner pass, because
+        // an expunged sled is one that has been removed from the rack either
+        // physically or via trust-quorum.
+        if sled_details.policy == SledPolicy::Expunged {
+            for (_, disk) in sled_details.resources.all_disks(DiskFilter::All) {
                 editor
-                    .expunge_disk(&disk_id)
+                    .decommission_disk(&disk.disk_id)
                     .map_err(|err| Error::SledEditError { sled_id, err })?;
             }
+        } else {
+            // The sled is not expunged. We have to see if the inventory
+            // reflects the parent blueprint disk generation. If it does
+            // then we mark any expunged disks decommissioned.
+            let seen_generation = self
+                .collection
+                .sled_agents
+                .get(&sled_id)
+                .map(|sa| sa.omicron_physical_disks_generation);
+
+            // Do we have any expunged disks?
+            let expunged_disks: Vec<_> = editor
+                .disks(BlueprintPhysicalDiskDisposition::is_expunged)
+                .cloned()
+                .collect();
+            for disk in expunged_disks {
+                if let BlueprintPhysicalDiskDisposition::Expunged {
+                    as_of_generation,
+                    ..
+                } = disk.disposition
+                {
+                    // Has the sled agent seen this disk's expungement yet as
+                    // reflected in inventory?
+                    if seen_generation >= Some(as_of_generation) {
```

Comment:
This is correct but potentially comparing `None >= Some(as_of_generation)` feels icky. Is there a way we could make `seen_generation` non-optional by the time we get here, since if we didn't see this sled in the inventory collection we can't do anything? Could we make the else on line 1137 an `else if let Some(seen_generation) = ...`?